controller/sync_strats/sync_instant.go

```diff
@@ -0,0 +1,528 @@
+package sync_strats
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"github.com/lucsky/cuid"
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/edge/build"
+	"github.com/openziti/edge/controller/env"
+	"github.com/openziti/edge/controller/handler_edge_ctrl"
+	"github.com/openziti/edge/controller/model"
+	"github.com/openziti/edge/controller/persistence"
+	"github.com/openziti/edge/pb/edge_ctrl_pb"
+	"github.com/openziti/fabric/controller/network"
+	"github.com/openziti/foundation/channel2"
+	"github.com/openziti/foundation/storage/ast"
+	"go.etcd.io/bbolt"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	RouterSyncStrategyInstant env.RouterSyncStrategyType = "instant"
+)
+
+var _ env.RouterSyncStrategy = &InstantStrategy{}
+
+// Original API Session synchronization implementation. Assumes that on connect, the router requires and instant
+// and full set of API Sessions
+type InstantStrategy struct {
+	InstantStrategyOptions
+
+	rtxMap *routerTxMap
+
+	helloHandler  channel2.ReceiveHandler
+	resyncHandler channel2.ReceiveHandler
+	ae            *env.AppEnv
+
+	helloOutQueue chan *routerTx
+	helloInQueue  chan *routerTx
+
+	stopHelloOut chan struct{}
+	stopHelloIn  chan struct{}
+}
+
+func (strategy *InstantStrategy) GetOnlineEdgeRouter(id string) (*model.EdgeRouter, env.RouterSyncStatus) {
+	rtx := strategy.rtxMap.Get(id)
+
+	if rtx != nil {
+		return rtx.EdgeRouter, rtx.Status
+	}
+
+	return nil, env.RouterSyncUnknown
+}
+
+func (strategy *InstantStrategy) Status(id string) env.RouterSyncStatus {
+	rtx := strategy.rtxMap.Get(id)
+	if rtx == nil {
+		return env.RouterSyncUnknown
+	}
+
+	return rtx.Status
+}
+
+func NewInstantStrategy(ae *env.AppEnv, options InstantStrategyOptions) *InstantStrategy {
+	if options.MaxOutstandingHellos <= 0 {
+		pfxlog.Logger().Panicf("MaxOutstandingHellos for InstantStrategy cannot be less than 1, got %d", options.MaxOutstandingHellos)
+	}
+
+	if options.MaxConcurrentSyncs <= 0 {
+		pfxlog.Logger().Panicf("MaxConcurrentSyncs for InstantStrategy cannot be less than 1, got %d", options.MaxConcurrentSyncs)
+	}
+
+	if options.RouterTxBufferSize < 0 {
+		pfxlog.Logger().Panicf("RouterTxBufferSize for InstantStrategy cannot be less than 0, got %d", options.MaxOutstandingHellos)
+	}
+
+	strategy := &InstantStrategy{
+		InstantStrategyOptions: options,
+		rtxMap: &routerTxMap{
+			internalMap: &sync.Map{},
+		},
+		ae:            ae,
+		helloOutQueue: make(chan *routerTx, 100),
+		helloInQueue:  make(chan *routerTx, 100),
+		stopHelloOut:  make(chan struct{}, 0),
+		stopHelloIn:   make(chan struct{}, 0),
+	}
+
+	strategy.helloHandler = handler_edge_ctrl.NewHelloHandler(ae, strategy.ReceiveHello)
+	strategy.resyncHandler = handler_edge_ctrl.NewResyncHandler(ae, strategy.ReceiveResync)
+
+	for i := int32(0); i < options.MaxOutstandingHellos; i++ {
+		go strategy.startHelloWorker()
+	}
+
+	for i := int32(0); i < options.MaxOutstandingHellos; i++ {
+		go strategy.startSynchronizeWorker()
+	}
```

Comment:
This uses two sets of workers to handle the initial hello sync and then API sessions/session sync. 

This should help alleviate issues when multiple edge routers are reconnecting at once. The edge routers sit and wait for the controller to respond to them at each step.